prediction_market_agent/agents/custom_agent.py

```diff
@@ -0,0 +1,119 @@
+        self.keys = utils.get_keys()
+        self.tools = {
+            "GoogleSearchTool": google_search,
+            "WebScrapingTool": tool_exception_handler(
+                map_exception_to_output={
+                    requests.exceptions.HTTPError: "Couldn't reach the URL. Don't scrap the same domain again."
+                }
+            )(web_scrape_structured_and_summarized),
+        }
+        self.system_prompt = f"""You are a research assistant meant to answer questions placed on a predictive market.
+
+To do that, you can't rely on your training data, but you have access to the following tools in order to answer question:
+
+1. GoogleSearchTool: Given a query, returns a list of URLs. Accepts arguments as a dictionary, with a single key "query".
+2. WebScrapingTool: Given an objective and url, returns the summarized content of the page. In the objective, ask closed-ended questions. Accepts arguments as a dictionary, with keys "objective" and "url".
+
+If you want to use the tool, simply return completion in form of a dictionary with the keys "tool_name" and "tool_params".
+
+If you used a tool in the past, its output will be available in a message in form of a dictionary with the keys "tool_name" and "tool_output".
+
+If you need to think about the question, return a completion in form of a dictionary with a single key "thinking" and a string value.
+
+If you aren't sure, feel free to continue research by using the tools and following URLs until you have all the information, unless you are told otherwise.
+
+If you want to answer, return a completion in form of a dictionary with a single key "answer" and a string value.
+"""
+        self.model = ChatOpenAIModel(
+            model="gpt-4",  # I tried gpt-3.5-turbo, but it fails at providing correct inputs to the tools.
+            system_prompt=self.system_prompt,
+            api_key=self.keys.openai,
+        )
+
+    def verbose_print(self, message: str) -> None:
+        if self.verbose:
+            print(f"{message}\n")
+
+    def run(self, market: str) -> bool:
+        cycle_count = 0
+        answer: Optional[str] = None
+        # Get the main objective prompt.
+        objective = utils.get_market_prompt(market)
+        # Keep track of history of messages.
+        messages = [objective]
+
+        while True:
+            cycle_count += 1
+            # If we reached the maximum number of cycles, ask to provide an answer based on available information.
+            if cycle_count == self.max_cycles - 1:
+                self.verbose_print(
+                    f"Agent: Reached {cycle_count} cycles, asking to provide answer now."
+                )
+                messages.append(
+                    "You have reached maximal message count, please provide answer now as best as you can."
+                )
+
+            # Because of the system prompt, completions are expected to be JSON-parseable.
+            completion_str = self.model.complete(messages)
```

Comment:
Interestingly, it works nicely without specifying other roles, just `system` and `user` here. But if we want to improve this model, the use of `assistant` and `tool` roles could help (at least in the case of OpenAI models).

Reply:
Interesting. Maybe it's happening by default under the hood if not specified?

Reply:
Inside of `ChatOpenAIModel.complete`, I'm automatically assigning the `user` role to these messages, so I don't think they re-write it. But after all, it will all just be merged into one prompt, so maybe the model isn't super-sensitive about it.